### ISSUE TITLE
Load network route table names from /etc/iproute2/rt_tables**

### ISSUE THREAD

**Issue by hachi** [systemd/systemd#35498]
### Component

systemd-networkd

### Is your feature request related to a problem? Please describe

I don't want to define my table names in multiple places, and the standard place for `ip` and related tools is /etc/iproute2/* files. There are a lot of tools that use `ip` and it's a little dangerous to make the same data be two places unsynchronized.

### Describe the solution you'd like

Parse /etc/iproute2/* files for settings 

### Describe alternatives you've considered

Could also make systemd maintain the iproute2 files somehow... not sure.

**Comment by hachi** [comment]
I see that this idea was discounted long ago in: https://github.com/systemd/systemd/pull/3132

but systemd is mature and popular enough now that the effort is probably worth it to be compatible with the sheer number of things that use iproute2

**Comment by poettering** [comment]
I am not a fan of the idea of consuming some other packages config files. Via their choice of path they made very clear it's a private iproute2 file, and not a generic system concept, hence I am not sure we should overreach there, introduce a new lifecycle and so on.

**Comment by hachi** [comment]
With incredible respect to you folks maintaining systemd... that file path was chosen back sometime near 1999. That means it has been the standard location for the definition of the route table names for 25 years now. Calling it a new lifecycle is a little myopic, and I'm pretty sure the format and path have not changed at all in that time.

Does systemd have a project goal of reducing user mistakes? I'm suggesting this because I accidentally numbered my route tables backwards when copying it in, and then spent many hours trying to figure out what wasn't working correctly.

I don't think the answer needs to be that you consume the file directly. But just saying 'no thanks' doesn't really help the compatibility with other things. There is also the fact that the iproute2 tools are still the standard tools for looking at the routing tables, even in a systemd managed system. ( https://www.freedesktop.org/software/systemd/man/latest/systemd.network.html#Table=1 )

I think the ultimate solution would be to allow table name lookup to happen in some way other than just networkd.conf. Is there any way to make it be another program via exec or bus communication that can do route table name lookups? I don't see any normal methods in the systemd.syntax side of things, but maybe I'm not clever enough.

A bonus goal would be to be able to check the mapping in both places, even if hard coded, and issue a warning which can be silenced if the table names do not match.

**Comment by yuwata** [comment]
Basically I like the idea.

**Comment by poettering** [comment]
> With incredible respect to you folks maintaining systemd... that file path was chosen back sometime near 1999. That means it has been the standard location for the definition of the route table names for 25 years now. Calling it a new lifecycle is a little myopic, and I'm pretty sure the format and path have not changed at all in that time.

My use of the word lifecycle refers to the fact that networkd is a long-running service that parses configuration and keeps in memory. hence, we'd read files whose semantics are defined by some other package, but we'd use them with quite different lifecycle logic than the original package (because iproute2 is *not* short-running but loads the files on every invocation fresh).

hence, I don't think it's so obvious to read this stuff: it's not owned by us, and we have to figure out what to do about the lifecycle. do we use inotify or mtimes to watch the stuff? but if we do, we also encode expected update logic for the file and that all becomes a mess.

if we own these files things are easy, because we can just define the semantics. If the files are a widely established "standard" already processed by a myriad of packages (e.g. /etc/resolv.conf) then we can figure things out too, because we know that noone can realistically change semantics anymore because there are so many stakeholders. But a private file of one package? i am not too convinced it's that easy.

in particular as the file format has exactly zero documentation afaics? there's some 3rd party docs, but iproute's documentation just mentions its existance, not its format afaics?

**Comment by poettering** [comment]
so what i think might be ok would be something like this: maybe introduce our own drop-in dir for routing table names (i.e {/etc,/run/,/usr/lib}/rttable.d/*.rttable, then define clear semantics for it. use a format we derive from iproute2's parser for this and document it properly. 

The idea would be then that we have clear semantics, but because the file format for iproute2 and our stuff would happen to be compatible (because we made them so), it would be up to people to then optionally symlink the files they maintain for iproute2 also into this generic directory, to sync them up. (such a symlink could even be create as part of the packaging of systemd).

And then maybe it would be worth contacting iproute2 maintainers to maybe consider supporting the generic drop-in dir too.

to me, such an approach would solve the "overreach" into iproute2's private files. and it would allow us to define clear semantics on lifecycle and file format, while providing enough compat for most people to suffice.

**Comment by poettering** [comment]
> so what i think might be ok would be something like this: maybe introduce our own drop-in dir for routing table names (i.e {/etc,/run/,/usr/lib}/rttable.d/*.rttable, then define clear semantics for it. use a format we derive from iproute2's parser for this and document it properly.

oh, and if we do this, i'd probably say we shouldn't do inotify/mtime, but instead require "systemctl reload systemd-networkd" for the files to be considered.

**Comment by hachi** [comment]
I'm sorry for my response on the term lifecycle then. You are absolutely correct about the difference in life of networkd vs iproute2 tools. 

I suggested the use of IPC to perform the lookup because it would keep it out of networkd memory, and if someone wants to use absurd numbers or dynamically allocated table names then it's possible, and kept outside the scope of networkd itself. To be clear that's only a thought, not a request from me.

You are also correct about the lacking definition of the particular file format, it has always been a defacto definition of format that just came from the default file shipped with the package: https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/tree/etc/iproute2

Would it be a reasonable suggestion that the search paths be defined in config? That would allow the laziness of the user to simply define the config to include whatever files they want, without symlinks, and also the future of putting the files wherever needed. I have no issue with /etc/rttables.d, but I also figure you don't need to make it be a thing defined by networkd if you don't want to.

The iproute2 method is to pick up `/etc/iproute2/rt_tables` and `/etc/iproute2/rt_tables.d/*.conf` but the manpage also says it will check `/usr/share/iproute2/rt_tables` and doesn't mention whether `/usr/share/iproute2/rt_tables.d` will also be included.

As for inotify/mtime I assumed `networkctl reload` would be the method to reload the config, which I assume is the same behavior as you're proposing. inotify/mtime introduces all sorts of headaches about removing entries when files are removed or changed, and I think that complexity is not reasonable.

**Comment by poettering** [comment]
> Would it be a reasonable suggestion that the search paths be defined in config?

No, we want clear interfaces, and that means fixed paths for things. It's fine if people redirect those paths elsewhere via symlinks or bind mounts, but the entrypoint should always be a fixed, well-known path.

**Comment by ChristianTacke** [comment]
An other option to consider:

Create a generator that reads the iproute2 table name config and creates networkd config drop-ins in `/run/`.

This would resemble what's happening for a bunch of other generators and would share all the lifecycle aspects with thtose?

**Comment by Aison0** [comment]
Why not doing it the other way round? Use networkd.conf to define the routing table mappings and then provide a `/run/systemd/network/stub-iproute2-rt.conf` which can be symlinked in /etc/iproute2/rt_tables.d/

Having the line
`RouteTable=wan:500 router:1000 management:1001 guest:1010`
inside networkd.conf and then have an autogenerated iproute2 stub file with content below would be simple.
```
500 wan
1000 router
1001 management
1010 guest
```

**Comment by hachi** [comment]
Did this ever get considered for implementation?

**Comment by Aison0** [comment]
I guess not

**Comment by hachi** [comment]
@poettering I've rechecked this now and I'm noticing that iproute2 is still the de facto tools to use for routing table interaction, but systemd still has no way to use the same table name definitions. Can this be considered for implementation?
